src/crimson/osd/ops_executer.h

```diff
@@ -233,25 +233,37 @@ auto OpsExecuter::with_effect_on_obc(
   return std::forward<MainFunc>(main_func)(ctx_ref);
 }
 
-template <typename Func>
-OpsExecuter::osd_op_errorator::future<> OpsExecuter::submit_changes(Func&& f) && {
+template <typename Func,
+          typename MutFunc>
+OpsExecuter::osd_op_errorator::future<> OpsExecuter::flush_changes(
+  Func&& func,
+  MutFunc&& mut_func) &&
+{
   assert(obc);
-  if (!osd_op_params) {
-    osd_op_params = osd_op_params_t();
+  const bool want_mutate = !txn.empty();
+  if (want_mutate) {
+    if (!osd_op_params) {
+      osd_op_params = osd_op_params_t();
+    }
+    osd_op_params->req = std::move(msg);
+    osd_op_params->at_version = pg.next_version();
+    osd_op_params->pg_trim_to = pg.get_pg_trim_to();
+    osd_op_params->min_last_complete_ondisk = pg.get_min_last_complete_ondisk();
+    osd_op_params->last_complete = pg.get_info().last_complete;
+    if (user_modify) {
+      osd_op_params->user_at_version = osd_op_params->at_version.version;
+    }
   }
-  osd_op_params->req = std::move(msg);
-  eversion_t at_version = pg.next_version();
-
-  osd_op_params->at_version = at_version;
-  osd_op_params->pg_trim_to = pg.get_pg_trim_to();
-  osd_op_params->min_last_complete_ondisk = pg.get_min_last_complete_ondisk();
-  osd_op_params->last_complete = pg.get_info().last_complete;
-  if (user_modify)
-    osd_op_params->user_at_version = at_version.version;
   if (__builtin_expect(op_effects.empty(), true)) {
-    return std::forward<Func>(f)(std::move(txn), std::move(obc), std::move(*osd_op_params));
+    return want_mutate ? std::forward<MutFunc>(mut_func)(std::move(txn),
+                                                         std::move(obc),
+                                                         std::move(*osd_op_params))
+                       : std::forward<Func>(func)(std::move(obc));
   }
-  return std::forward<Func>(f)(std::move(txn), std::move(obc), std::move(*osd_op_params)).safe_then([this] {
+  return (want_mutate ? std::forward<MutFunc>(mut_func)(std::move(txn),
```

Comment:
nit, i'd suggest move this `return` clause into a `else` block for better readability.